### PR TITLE
Fix invalid index analysis for date related keys

### LINKED ISSUE #44504: when i use"LIKE" in version 22.3.2.1 ,there is something wrong with "Cannot parse string to Date"
when i use "LIKE" in version 20.3.3.6 is all right,but in 22.3.2.1 something was wrong
```
clickhouse-180 :) show create table act.goodsdeliverflow_local

SHOW CREATE TABLE act.goodsdeliverflow_local

Query id: 846831d1-a00c-4ae4-bbe1-8322dd118141

┌─statement──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE act.goodsdeliverflow_local
(

     .......
    `dtEventTime` String,
     .......
)
ENGINE = MergeTree
PARTITION BY toYYYYMM(toDate(dtEventTime))
ORDER BY (dtEventTime, localeCountryCode, PlatID)
SETTINGS index_granularity = 8192 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 rows in set. Elapsed: 0.003 sec.

clickhouse-180 :) select * from act.goodsdeliverflow_local where dtEventTime LIKE '2022-10-01%'

SELECT *
FROM act.goodsdeliverflow_local
WHERE dtEventTime LIKE '2022-10-01%'

Query id: f335868e-edf1-42c3-a7c1-36d774096db6


0 rows in set. Elapsed: 0.007 sec.

Received exception from server (version 22.3.2):
Code: 6. DB::Exception: Received from localhost:9099. DB::Exception: Cannot parse string '2022-10-01%' as Date: syntax error at position 10 (parsed just '2022-10-01'). (CANNOT_PARSE_TEXT)

clickhouse-180 :)
````

why i use "like" for a `String` column,but the error messsage told me `Cannot parse string to Date`
Is it because the column is the partition key

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix some broken cases of index analysis when the table contains date/datetime related partition keys or primary keys. This fixes #44504.

### CODE REVIEW

File: src/Storages/MergeTree/KeyCondition.cpp [hunk 1]
```diff
@@ -1026,10 +1039,23 @@ bool KeyCondition::transformConstantWithValidFunctions(
                     if (func->type != ActionsDAG::ActionType::FUNCTION)
                         continue;
 
+                    const auto & func_name = func->function_base->getName();
+                    auto func_base = func->function_base;
+                    if (date_time_parsing_functions.contains(func_name))
```
[the remark this answers is not in the document]
Comment (amosbird) [reply src/Storages/MergeTree/KeyCondition.cpp:1044]: I don't think it's needed. I cannot think of a valid query that will have a `const_type_ which breaks `-OrNull` variants.

[the remark this answers is not in the document]
Comment (canhld94) [reply src/Storages/MergeTree/KeyCondition.cpp:1044]: For example, table has partition expression `toDate(ts)` with `ts` is `DateTime` and predicate is `ts >= toDateTime('2023-01-01 00:00:00')` then '2023-01-01 00:00:00' wii be casted to `DateTime` first before transformation. `toDateOrNull` will throw exception if argument type is `DateTime`. The same for all conversion function from string.

[the remark this answers is not in the document]
Comment (amosbird) [reply src/Storages/MergeTree/KeyCondition.cpp:1044]: Interesting. I didn't know that `-OrNull` variants only accept string)

**Review by alexey-milovidov — APPROVED** [review]
There was a major offense: missing comments.
I've added a comment according to my understanding of the changes.

**Pull request merged**